src/buffer.cc

```diff
@@ -204,6 +205,82 @@ Result Buffer::CompareRMSE(Buffer* buffer, float tolerance) const {
   return {};
 }
 
+std::vector<double> Buffer::GetHistogramForChannel(uint32_t channel,
+                                                   uint32_t num_bins) const {
+  std::vector<double> bins(num_bins, 0.0);
+  auto* buf_ptr = GetValues<uint8_t>();
+  const auto& segments = format_->GetSegments();
+  const auto comp = segments[channel].GetComponent();
+  auto num_channels = format_->ValuesPerElement();
+  buf_ptr += channel * comp->SizeInBytes();
+  for (size_t i = 0; i < ElementCount(); ++i) {
+    const auto bin = *reinterpret_cast<const uint8_t*>(buf_ptr);
+    bins[bin]++;
+    // Assumes unpacked format where all channels have the same bit size
+    buf_ptr += comp->SizeInBytes() * num_channels;
+  }
+
+  return bins;
+}
+
+Result Buffer::CompareHistogramEMD(Buffer* buffer, float tolerance) const {
+  auto result = CheckCompability(buffer);
+  if (!result.IsSuccess())
+    return result;
+
+  const int num_bins = 256;
+  auto num_channels = format_->ValuesPerElement();
+
+  for (auto segment : format_->GetSegments()) {
+    if (!segment.GetComponent()->IsUint8() || num_channels != 4)
+      return Result(
+          "EMD comparison only supports 8bit unorm format with four channels.");
+  }
+
+  std::vector<std::vector<double>> histograms[2];
+  for (uint32_t c = 0; c < num_channels; ++c) {
+    histograms[0].push_back(GetHistogramForChannel(c, num_bins));
+    histograms[1].push_back(buffer->GetHistogramForChannel(c, num_bins));
+  }
+
+  // Normalize histograms
+  for (uint32_t i = 0; i < 2; ++i)
+    for (uint32_t c = 0; c < num_channels; ++c) {
+      double total = 0;
+      for (auto value : histograms[i][c])
+        total += value;
+      for (auto& value : histograms[i][c])
+        value /= total;
+    }
+
+  double max_emd = 0;
+
+  for (uint32_t c = 0; c < num_channels; ++c) {
+    std::vector<double> diffs(histograms[0][c].size());
+    for (size_t i = 0; i < histograms[0][c].size(); ++i)
```

Comment:
This should probably work off the segments, and use CalculateDiff above for consistency.

Reply:
Wouldn't that complicate things? The histogram format is different from the original pixel format, so we would need to create new format object for this to work. And add pointer iteration instead of one single for loop.

Reply:
You don't need a new format. CalculateDiff just takes a single segment which in this case is just your Uint, 8 bits. Or, loop over the segments from the buffer format and use those.